kyo-core/shared/src/main/scala/kyo/Channel.scala

```diff
@@ -286,234 +293,418 @@ object Channel:
+                flush()
+                promise
+            end takeFiber
+        end BaseUnsafe
+
+        final class ZeroCapacityUnsafe[A](using initFrame: Frame, allow: AllowUnsafe) extends BaseUnsafe[A]:
+            val isClosed = new java.util.concurrent.atomic.AtomicBoolean()
+
+            private val closedResult                     = Result.fail(Closed("Channel", initFrame, "zero-capacity"))
+            def asResult[A](value: A): Result[Closed, A] = if isClosed.get() then closedResult else Result.succeed(value)
+
+            def capacity = 0
+
+            def size()(using AllowUnsafe) = asResult(0)
+
+            def offer(value: A)(using AllowUnsafe) =
+                Maybe(takes.poll()) match
+                    case Absent =>
+                        asResult(puts.offer(Put.Value(value, Promise.Unsafe.init[Closed, Unit]())))
+                    case Present(takePromise) =>
+                        if takePromise.complete(Result.succeed(value)) then asResult(true)
+                        else offer(value)
+
+            def offerAll(values: Seq[A])(using AllowUnsafe): Result[Closed, Chunk[A]] =
+                @tailrec def loop(currentChunk: Chunk[A]): Result[Closed, Chunk[A]] =
+                    currentChunk.headMaybe match
+                        case Absent =>
+                            asResult(Chunk.empty)
+                        case Present(value) =>
+                            Maybe(takes.poll()) match
+                                case Absent =>
+                                    asResult {
+                                        if puts.offer(Put.Batch(currentChunk, Promise.Unsafe.init[Closed, Unit]())) then Chunk.empty
+                                        else currentChunk
+                                    }
+                                case Present(takePromise) =>
+                                    if takePromise.complete(Result.succeed(value)) then loop(currentChunk.dropLeft(1))
+                                    else loop(currentChunk)
+
+                loop(Chunk.from(values))
+            end offerAll
+
+            def poll()(using AllowUnsafe) =
+                asResult {
+                    Maybe(puts.poll()) match
+                        case Absent =>
+                            Absent
+                        case Present(Put.Value(value, promise)) =>
+                            discard(promise.complete(Result.unit))
+                            flush()
+                            Present(value)
+                        case Present(Put.Batch(batch, promise)) =>
+                            val result = batch.headMaybe match
+                                case Absent =>
+                                    discard(promise.complete(Result.unit))
+                                    Absent
+                                case Present(value) =>
+                                    if batch.tail.nonEmpty then discard(puts.offer(Put.Batch(batch.tail, promise)))
```

Comment:
This can break fairness since the batched put goes back to the end of the queue. It's pre-existing so we could follow up separately